### PR TITLE
[SPARK-7899][PYSPARK] Fix Python 3 pyspark/sql/types module conflict

### PR DESCRIPTION
**Pull request by ghost**
This PR makes the types module in `pyspark/sql/types` work with pylint static analysis by removing the dynamic naming of the `pyspark/sql/_types` module to `pyspark/sql/types`.

Tests are now loaded using `$PYSPARK_DRIVER_PYTHON -m module` rather than `$PYSPARK_DRIVER_PYTHON module.py`. The old method adds the location of `module.py` to `sys.path`, so this change prevents accidental use of relative paths in Python.

### CODE REVIEW

File: python/run-tests [hunk 1]
```diff
@@ -57,54 +57,54 @@ function run_test() {
 
 function run_core_tests() {
     echo "Run core tests ..."
-    run_test "pyspark/rdd.py"
-    run_test "pyspark/context.py"
-    run_test "pyspark/conf.py"
-    PYSPARK_DOC_TEST=1 run_test "pyspark/broadcast.py"
```
Comment (justinuang) [inline python/run-tests]: Why did you remove the PYSPARK_DOC_TEST env var?

Comment (mnazario) [inline python/run-tests]: I removed PYSPARK_DOC_TEST to make the doctests consistent with the other pyspark tests since I was changing all of the other test names to modules instead of filepaths.

The two doctests now have the appropriate main function which runs the doctests for that module.

**Comment by JoshRosen** [comment]
Jenkins, this is ok to test.

File: python/run-tests (hunk 1 above)
Comment (justinuang) [inline python/run-tests]: alright, sounds good!

**Comment by davies** [comment]
This looks good to me, could you also update pyspark/mllib/rand.py to random.py? (and remove the hack).

One downside of this is that we can not run single test suite by spark-submit:
```
$ bin/spark-submit python/pyspark/sql/dataframe.py
```
Instead, we should do:
```
$ PYTHON_PATH=$HOME/python:$HOME/python/lib/py4j-0.8.2.1-src.zip python -m pyspark.sql.dataframe
```

**Comment by mnazario** [comment]
I'll remove the hack. I don't really think of the spark-submit change as a downside since it's the proper way it should be run for now.
Once this stuff is moved to an actual test running like nose, then you'll be able to selectively choose which tests need to be run that way.

**Comment by davies** [comment]
LGTM

**Comment by JoshRosen** [comment]
@mnazario, since this was motivated by running `pylint` checks, I'd be very curious to know whether pylint uncovered any bugs or dodgy code.  Do you have any results that you can share?

**Comment by mnazario** [comment]
We were running `pylint` on our code which was depending on pyspark. The dynamic module naming  confused pylint.
